### PR TITLE
A user can search for available bedspaces

### PR DESCRIPTION
**Pull request by bryangaledxw**
# Changes in this PR

This PR adds a search bedspaces feature to the service. It does inital groundwork and tidying, before adding a client, service, and controller, then E2E testing once the feature is reachable through natural navigation

This is a first iteration of the feature. Areas that may improve soon:
- Layout. I tried positioning the date and the duration side-by-side, but this makes the difference between date fields and regular text input fields very obvious, particularly when it comes to error handling. This may be surmountable, but it's sufficiently fiddly to leave to one side for now
- PDU filtering. The user currently needs to select a PDU from a list of all PDUs. We should filter this list to show only the PDUs relevant to the user's region, and possibly make PDU optional
- Journey. Currently the user can only return to the search page from the bedspace page by using the browser back button. We should have a link that will take the user back

### Before
localhost_3000_ (5)

### After
localhost_3000_ (4)
localhost_3000_find-a-bedspace_startDate-day=1 startDate-month=1 startDate-year=2022 durationDays=4 probationDeliveryUnit=Birmingham+North%2C+East+and+Solihull
localhost_3000_find-a-bedspace_startDate-day=1 startDate-month=1 startDate-year=2022 durationDays=4 probationDeliveryUnit=Barking%2C+Dagenham+and+Havering

# Release checklist

As part of our continuous deployment strategy we must ensure that this work is
ready to be released at any point. Before merging to `main` we must first
confirm:

## Pre merge checklist

- [x] Have all changes to any dependencies been deployed to both `preprod` and
    `production` in a backwards compatible way? (This includes our API,
    infrastructure, third-party integrations and any secrets or environment variables)
- [x] Has a required data migration been included in this change or been done in
    advance?

## Post merge checklist

Once we've merged we now need to release this through to production before
considering it done.

[Find the build-and-deploy job in CircleCI](https://app.circleci.com/pipelines/github/ministryofjustice/hmpps-temporary-accommodation-ui)
and work through the following steps:

- [x] Has the product manager asked to provide approval for this feature?
  - [x] Has the product manager approved?
- [x] Manually approve release to preprod
- [ ] Verify change released to preprod
- [x] Manually approve release to prod
- [ ] Verify change released to production

Should a release fail at any step, you as the author should now lead the work to
fix it as soon as possible.

### CODE REVIEW

**Comment by bryangaledxw** [comment]
[A user can search for available bedspaces](https://trello.com/c/pp4X8gnt/914-a-user-can-search-for-available-bedspaces)

File: e2e/tests/stepDefinitions/manage.ts [hunk 1]
```diff
@@ -11,3 +12,7 @@ Given('I am logged in', () => {
 
   cy.get('.govuk-button').contains('Sign in').click()
 })
+
+Given('I return to the dashboard', () => {
+  DashboardPage.clickDashboardLink()
```
Comment (patrickjfl) [inline e2e/tests/stepDefinitions/manage.ts]: I'm confused by this. Why is this a method on DashboardPage? Isn't that the only page where we don't need it?

File: cypress_shared/pages/page.ts [hunk 2]
```diff
@@ -60,6 +60,54 @@ export default abstract class Page extends Component {
     cy.get(`select[id="${id}"]`).children('option').contains(exact(contents)).should('be.selected')
   }
 
+  shouldShowDateInputsByLegend(legend: string, date: string): void {
```
Comment (patrickjfl) [inline cypress_shared/pages/page.ts:67]: nice

File: e2e/tests/bedspaceSearch.feature [hunk 3]
```diff
@@ -0,0 +1,17 @@
...
+        And I create a bedspace with all necessary details
+        And I return to the dashboard
+        And I'm searching for a bedspace
+        And I search for a bedspace
+        Then I should see the bedspace search results
+
+    Scenario: Showing bedspace search errors
+        Given I'm searching for a bedspace
+        And I attempt to search for a bedspace with required details missing
+        Then I should see a list of the problems encountered searching for a bedspace
```
Comment (patrickjfl) [inline e2e/tests/bedspaceSearch.feature]: no newline at end of file err

**Review by patrickjfl — APPROVED** [review]
LGTM 👍

File: cypress_shared/pages/page.ts (hunk 2 above)
Comment (bryangaledxw) [reply cypress_shared/pages/page.ts:67]: I was put off doing this before as I don't have time to go through everything and make it super consistent, but I think I need to just get this in and we start using this approach where we can

**Pull request merged**